includes/class-sensei-course-progression-settings.php

```diff
@@ -0,0 +1,107 @@
+<?php
+/**
+ * File containing the Sensei_Course_Progression_Settings class.
+ *
+ * @package sensei-lms
+ * @since 4.0.0
+ */
+
+if ( ! defined( 'ABSPATH' ) ) {
+	exit; // Exit if accessed directly.
+}
+
+/**
+ * Adds the course progression settings to the course.
+ *
+ * @since 4.0.0
+ */
+class Sensei_Course_Progression_Settings {
+
+	const VIDEO_COURSE_AUTOCOMPLETE = '_video_course_autocomplete';
+	const VIDEO_COURSE_AUTOPAUSE    = '_video_course_autopause';
+	const VIDEO_COURSE_REQUIRED     = '_video_course_required';
+
+	/**
+	 * Instance of class.
+	 *
+	 * @var self
+	 */
+	private static $instance;
+
+	/**
+	 * Sensei_Course_Progression_Settings constructor. Prevents other instances from being created outside of `self::instance()`.
+	 */
+	private function __construct() {
+	}
+
+	/**
+	 * Fetches an instance of the class.
+	 *
+	 * @return self
+	 */
+	public static function instance() {
+		if ( ! self::$instance ) {
+			self::$instance = new self();
+		}
+
+		return self::$instance;
+	}
+
+	/**
+	 * Initializes the Video-Based Course Progression.
+	 *
+	 * @param Sensei_Main $sensei Sensei object.
+	 */
+	public function init( $sensei ) {
+		add_action( 'admin_enqueue_scripts', [ $this, 'add_feature_flag_inline_script' ] );
+
+		if ( ! $sensei->feature_flags->is_enabled( 'video_based_course_progression' ) ) {
+			// As soon this feature flag check is removed, the `$sensei` argument can also be removed.
+			return;
+		}
+
+		add_action( 'init', [ $this, 'register_post_meta' ] );
+	}
+
+	/**
+	 * Add feature flag inline script.
+	 *
+	 * @access private
+	 */
+	public function add_feature_flag_inline_script() {
+		$screen  = get_current_screen();
+		$enabled = Sensei()->feature_flags->is_enabled( 'video_based_course_progression' ) ? 'true' : 'false';
+
+		if ( 'course' === $screen->id ) {
+			wp_add_inline_script( 'sensei-admin-course-edit', 'window.senseiVideoCourseProgressionFeatureFlagEnabled = ' . $enabled, 'before' );
+		}
+	}
+
+	/**
+	 * Register post meta.
+	 *
+	 * @access private
+	 */
+	public function register_post_meta() {
+		$settings = [
+			self::VIDEO_COURSE_AUTOCOMPLETE,
+			self::VIDEO_COURSE_AUTOPAUSE,
+			self::VIDEO_COURSE_REQUIRED,
+		];
+		foreach ( $settings as $setting ) {
+			register_post_meta(
+				'course',
+				$setting,
+				[
+					'show_in_rest'  => true,
+					'single'        => true,
+					'type'          => 'boolean',
+					'default'       => false,
+					'auth_callback' => function ( $allowed, $meta_key, $post_id ) {
+						return current_user_can( 'edit_post', $post_id );
```

Comment:
It seems more appropriate to use `edit_course` here since someone could theoretically be allowed to edit posts but not courses. However, I'm not sure that the teacher role [gets that capability](https://github.com/Automattic/sensei/blob/e93d28f424f4a493d9b3318e92b7fc922c69ae78/includes/class-sensei-teacher.php#L165). Maybe we'd have to use `current_user_can( 'edit_courses' );` instead?

Whatever the solution ends up being, a teacher should only be able to edit these settings for their own courses, but not for a course created by somebody else.